src/dev/mcp23017.rs

```diff
@@ -0,0 +1,285 @@
+    pub gpb2: crate::Pin<'a, crate::mode::Input, M>,
+    pub gpb3: crate::Pin<'a, crate::mode::Input, M>,
+    pub gpb4: crate::Pin<'a, crate::mode::Input, M>,
+    pub gpb5: crate::Pin<'a, crate::mode::Input, M>,
+    pub gpb6: crate::Pin<'a, crate::mode::Input, M>,
+    pub gpb7: crate::Pin<'a, crate::mode::Input, M>,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Regs {
+    // N.B.: These values are for BANK=0, which is the reset state of
+    // the chip (and this driver does not change).
+    //
+    // For all registers, the reset value is 0x00, except for
+    // IODIR{A,B} which are 0xFF (making all pins inputs) at reset.
+    //
+    // IODIR: input/output direction: 0=output; 1=input
+    // IPOL: input polarity: 0=register values match input pins; 1=opposite
+    // GPINTEN: interrupt-on-change: 0=disable; 1=enable
+    // DEFVAL: default values for interrupt-on-change
+    // INTCON: interrupt-on-change config: 0=compare to previous pin value;
+    //   1=compare to corresponding bit in DEFVAL
+    // IOCON: configuration register
+    // - Pin 7: BANK (which driver assumes stays 0)
+    // - Pin 6: MIRROR: if enabled, INT{A,B} are logically ORed; an interrupt on either
+    //          port will cause both pins to activate
+    // - Pin 5: SEQOP: controls the incrementing function of the address pointer
+    // - Pin 4: DISSLW: disables slew rate control on SDA
+    // - Pin 3: HAEN: no effect on MCP23017
+    // - Pin 2: ODR: interrupt pins are 0=active-driver outputs (INTPOL sets polarity)
+    //          or 1=open-drain outputs (overrides INTPOL)
+    // - Pin 1: INTPOL: interrupt pin is 0=active-low or 1=active-high
+    // - Pin 0: unused
+    // GPPU: GPIO pull-ups: enables weak internal pull-ups on each pin (when configured
+    //   as an input)
+    // INTF: interrupt flags: 0=no interrupt pending; 1=corresponding pin caused interrupt
+    // INTCAP: interrupt captured value: reflects value of each pin at the time that they
+    //   caused an interrupt
+    // GPIO: reflects logic level on pins
+    // OLAT: output latches: sets state for pins configured as outputs
+    IODIRA = 0x00,
+    IPOLA = 0x02,
+    GPINTENA = 0x04,
+    DEFVALA = 0x06,
+    INTCONA = 0x08,
+    IOCONA = 0x0a,
+    GPPUA = 0x0c,
+    INTFA = 0x0e,
+    INTCAPA = 0x10,
+    GPIOA = 0x12,
+    OLATA = 0x14,
+    IODIRB = 0x01,
+    IPOLB = 0x03,
+    GPINTENB = 0x05,
+    DEFVALB = 0x07,
+    INTCONB = 0x09,
+    IOCONB = 0x0b,
+    GPPUB = 0x0d,
+    INTFB = 0x0f,
+    INTCAPB = 0x11,
+    GPIOB = 0x13,
+    OLATB = 0x15,
+}
+
+impl From<Regs> for u8 {
+    fn from(r: Regs) -> u8 {
+        r as u8
+    }
+}
+
+pub struct Driver<I2C> {
+    i2c: I2C,
+    out: u16,
+    addr: u8,
+}
+
+impl<I2C> Driver<I2C> {
+    pub fn new(i2c: I2C, a0: bool, a1: bool, a2: bool) -> Self {
+        let addr = 0x20 | ((a2 as u8) << 2) | ((a1 as u8) << 1) | (a0 as u8);
+        Self {
+            i2c,
+            out: 0xffff,
```

Comment:
Hm: this might not be correct, though the driver works in my test program.  GPIO{A,B} are 0x0000 after reset.